src/components/AuthForm/AuthForm.tsx

```diff
@@ -0,0 +1,239 @@
+import {
+  FC,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from 'react';
+
+import { useNavigate } from 'react-router-dom';
+import classNames from 'classnames';
+import { createUser, getUserByEmail } from '../../api/users';
+import { User } from '../../types/User';
+import { AuthContext } from '../../context/AuthContext';
+import { ErrorNotification } from '../ErrorNotification';
+
+export const AuthForm: FC = () => {
+  const [email, setEmail] = useState('');
+  const [name, setName] = useState('');
+  const [needToRegister, setNeedToRegister] = useState(false);
+
+  const [loading, setLoading] = useState(false);
+  const [hasError, setHasError] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const { setUser } = useContext(AuthContext);
+  const navigate = useNavigate();
+
+  const saveUser = (user: User) => {
+    localStorage.setItem('user', JSON.stringify(user));
+    setUser(user);
+  };
+
+  const generateError = useCallback((message: string) => {
+    setErrorMessage(message);
+    setHasError(true);
+  }, []);
+
+  useEffect(() => {
+    const userData = localStorage.getItem('user');
+
+    if (!userData) {
+      return;
+    }
+
+    try {
+      const user = JSON.parse(userData) as User;
+
+      setUser(user);
+    } catch (error) {
+      generateError('Something went wrong!');
+    }
+  }, []);
+
+  const loadUser = async () => {
+    const user = await getUserByEmail(email);
+
+    if (user) {
+      saveUser(user);
+      navigate('/todos');
+    } else {
+      setNeedToRegister(true);
+    }
+  };
+
+  const registerUser = () => {
+    return createUser({ name, email })
+      .then(saveUser)
+      .then(() => navigate('/todos'));
+  };
+
+  const isEmailValid = () => {
+    const emailPattern = /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(.\w{2,3})+$/;
+
+    if (!email) {
+      generateError('Email can\'t be empty!');
+
+      return false;
+    }
+
+    if (!emailPattern.test(email)) {
+      generateError('Invalid email!');
+
+      return false;
+    }
+
+    return true;
+  };
+
+  const isNameValid = () => {
+    if (!name && needToRegister) {
+      generateError('Name can\'t be empty!');
+
+      return false;
+    }
+
+    if (name.length < 3 && needToRegister) {
+      generateError('The name must be at least 3 letters!');
+
+      return false;
+    }
+
+    return true;
+  };
+
+  const handleSubmit = async (event: React.FormEvent) => {
+    event.preventDefault();
+
+    if (!isEmailValid() || !isNameValid()) {
+      return;
+    }
+
+    setErrorMessage('');
+    setLoading(true);
+
+    try {
+      if (needToRegister) {
+        await registerUser();
+      } else {
+        await loadUser();
+      }
+    } catch (error) {
+      generateError('Something went wrtong');
```

Comment:
Provide more reliable information about errors, so that it is better and faster to find and fix them later
```suggestion
      generateError(`Something went wrong with ${error.message}`);
```